Homework 3/src/SortedLinkedList.java

```diff
@@ -0,0 +1,190 @@
+            } else {
+                previous.next = newNode;
+            }
+            return;
+        }
+
+        // Goes after current node; if node is last, append
+        if (node.next == null) {
+            Node newNode = new Node();
+            newNode.data = toAdd;
+            node.next = newNode;
+        } else {
+            previous = node;
+            addHelper(node.next, toAdd, previous);
+        }
+    }
+
+    /**
+     * Checks the size (number of data items) of the list.
+     * @return the size of the list
+     */
+    @Override
+    public int size() {
+        if (head == null) return 0;
+        return sizeHelper(head, 0);
+    }
+
+    private int sizeHelper(Node temp, int count) {
+        count++;
+        if (temp.next == null) return count;
+        return sizeHelper(temp.next, count);
+    }
+
+    /**
+     * Displays the values of the list.
+     */
+    @Override
+    public void display() {
+        if (head == null) System.out.println("[]");
+        displayHelper(head);
+    }
+
+    private void displayHelper(Node temp) {
+        if (temp == head) System.out.print("[");
+        System.out.print(temp.data);
+        if (temp.next != null) {
+            System.out.print(", ");
+            displayHelper(temp.next);
+        } else {
+            System.out.println("]");
+        }
+    }
+
+    /**
+     * Returns true if the key value is in the list.
+     * @param key String key to search
+     * @return true if found, false if not found
+     */
+    @Override
+    public boolean contains(String key) {
+        if (head == null || key == null) return false;
+        return containsHelper(head.next, key);
+    }
+
+    private boolean containsHelper(Node temp, String key) {
+        if (temp.data.equals(key)) return true;
+        if (temp.next == null) return false;
+        return containsHelper(temp.next, key);
+    }
+
+    /**
+     * Returns true is the list is empty.
+     * @return true if it is empty, false if it is not empty
+     */
+    @Override
+    public boolean isEmpty() {
+        return head == null;
+    }
+
+    /**
+     * Removes and returns the first String object of the list.
+     * @return String object that is removed. If the list is empty, returns null
+     */
+    @Override
+    public String removeFirst() {
+        if (head == null) return null;
+
+        String data = head.data;
+        if (head.next == null) {
+            head = null;
+        } else {
+            head = head.next;
+        }
+
+        return data;
+    }
+
+    /**
+     * Removes and returns String object at the specified index.
+     * @param index index to remove String object
+     * @return String object that is removed
+     * @throws RuntimeException for invalid index value (index < 0 || index >= size())
+     */
+    @Override
+    public String removeAt(int index) {
+        if (head == null || index < 0) throw new RuntimeException("Invalid index value.");
```

Comment:
If you had a constant time size method, could you use it here to simplify your validity check for the index?

Reply:
A constant time size Method would ensure There is no significant performance Impact for using it here